File: htdocs/database/location.php
<?php 
// HANDLER
require_once "../modules/.php";

// IMPORTS
module("functional", "check");
module("functional", "database");
module("functional", "post");

// SIGNAL
signal("functional");

// RESPONSE SET TO JSON FROM ANYONE
header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');

// CHECK ARGUMENTS
check($PST, "LANG");
check($PST, "ID");
check($PST, "NODE");
check($PST, "CLUSTER");

// CHECK ARGUMENT RELATIONSHIPS
if (!(isx($PST, "LANG") and (isx($PST, "ID") xor isx($PST, "NODE") xor isx($PST, "CLUSTER")))) {
    throw new incorrectArgumentsError();
}

// TYPES OF QUERIES
if (isx($PST, "ID")) {
    $problemsQuery = database_request(
        "SELECT node, ? FROM problems WHERE id = ?",
        [
            "name_" . $PST["LANG"],
            $PST["ID"]
        ]
    )->fetch_assoc();
    $nodesQuery = database_request(
        "SELECT cluster, ? FROM nodes WHERE node = ?",
        [
            "name_" . $PST["LANG"],
            $problemsQuery["node"]
        ]
    )->fetch_assoc();
    $clustersQuery = database_request(
        "SELECT tree, ? FROM clusters WHERE cluster = ?",
        [
            "name_" . $PST["LANG"],
            $nodesQuery["cluster"]
        ]
    )->fetch_assoc();
    $treesQuery = database_request(
        "SELECT ? FROM trees WHERE tree = ?",
        [
            "name_" . $PST["LANG"],
            $clustersQuery["tree"]
        ]
    )->fetch_assoc();
    echo json_encode([
        "id" => [
            "value" => $PST["ID"],
            "name" => $problemsQuery["name_" . $PST["LANG"]]
        ],
        "node" => [
            "value" => $problemsQuery["node"],
            "name" => $nodesQuery["name_" . $PST["LANG"]]
        ],
        "cluster" => [
            "value" => $nodesQuery["cluster"],
            "name" => $clustersQuery["name_" . $PST["LANG"]]
        ],
        "tree" => [
            "value" => $clustersQuery["tree"],
            "name" => $treesQuery["name_" . $PST["LANG"]]
        ]
    ]);
} elseif (isx($PST, "NODE")) {
    $nodesQuery = database_request(
        "SELECT cluster, ? FROM nodes WHERE node = ?",
        [
            "name_" . $PST["LANG"],
            $PST["NODE"]
        ]
    )->fetch_assoc();
    $clustersQuery = database_request(
        "SELECT tree, name_" . $PST["LANG"] . " FROM clusters WHERE cluster = '" . $nodesQuery["cluster"] . "'",
        "SELECT tree, ? FROM clusters WHERE cluster = ?",
        [
            "name_" . $PST["LANG"],
            $nodesQuery["cluster"]
        ]
    )->fetch_assoc();
    $treesQuery = database_request(
        "SELECT ? FROM trees WHERE tree = ?",
        [
            "name_" . $PST["LANG"],
            $clustersQuery["tree"]
        ]
    )->fetch_assoc();
    echo json_encode([
        "id" => [
            "value" => null,
            "name" => null
        ],
        "node" => [
            "value" => $PST["NODE"],
            "name" => $nodesQuery["name_" . $PST["LANG"]]
        ],
        "cluster" => [
            "value" => $nodesQuery["cluster"],
            "name" => $clustersQuery["name_" . $PST["LANG"]]
        ],
        "tree" => [
            "value" => $clustersQuery["tree"],
            "name" => $treesQuery["name_" . $PST["LANG"]]
        ]
    ]);
} else {
    $clustersQuery = database_request(
        "SELECT tree, ? FROM clusters WHERE cluster = ?",
        [
            "name_" . $PST["LANG"],
            $PST["CLUSTER"]
        ]
    )->fetch_assoc();
    $treesQuery = database_request(
        "SELECT ? FROM trees WHERE tree = ?",
        [
            "name_" . $PST["LANG"],
            $clustersQuery["tree"]
        ]
    )->fetch_assoc();
    echo json_encode([
        "id" => [
            "value" => null,
            "name" => null
        ],
        "node" => [
            "value" => null,
            "name" => null
        ],
        "cluster" => [
            "value" => $PST["CLUSTER"],
            "name" => $clustersQuery["name_" . $PST["LANG"]]
        ],
        "tree" => [
            "value" => $clustersQuery["tree"],
            "name" => $treesQuery["name_" . $PST["LANG"]]
        ]
    ]);
}
?>

File: htdocs/api/location.py
#
#   INIT
#

# INIT -> HANDLER
import os
import sys
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
from handler import *

# INIT -> EXTENSIONS
from extensions.database import *
from extensions.post import *
from extensions.response import *


#
#   FUNCTION
#

# FUNCTION -> DECLARATION
@csrf_exempt
def output(request: object) -> object:
    # FUNCTION -> SUPERGLOBALS
    SUG.THR.REQ = request
    SUG.THR.SID = SUG.THR.REQ.COOKIES.get('session')
    
    # FUNCTION -> ACTIVATION
    database_init()
    post_init()
    response_init()
    
    # FUNCTION -> ARGUMENT CHECKS
    check("LANG")
    check("PROBLEM")
    check("NODE")
    check("CLUSTER")
    
    # FUNCTION -> ARGUMENT RELATIONSHIP
    if not (isx("LANG") and (isx("PROBLEM") ^ isx("NODE") ^ isx("CLUSTER"))):
        raise Error()

    # FUNCTION -> TYPES OF QUERIES
    if isx("PROBLEM"):
        problemsQuery = database_request(
            "SELECT node, ! FROM problems WHERE problem = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                SUG.THR.PST["PROBLEM"]
            ]
        )[0]
        nodesQuery = database_request(
            "SELECT cluster, ! FROM nodes WHERE node = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                problemsQuery['node']
            ]
        )[0]
        clustersQuery = database_request(
            "SELECT tree, ! FROM clusters WHERE cluster = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                nodesQuery['cluster']
            ]
        )[0]
        treesQuery = database_request(
            "SELECT ! FROM trees WHERE tree = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                clustersQuery['tree']
            ]
        )[0]
        problemValue = SUG.THR.PST["PROBLEM"]
        problemName = problemsQuery["name_" + SUG.THR.PST["LANG"]]
        nodeValue = problemsQuery['node']
        nodeName = nodesQuery["name_" + SUG.THR.PST["LANG"]]
        clusterValue = nodesQuery['cluster']
        clusterName = clustersQuery["name_" + SUG.THR.PST["LANG"]]
        treeValue = clustersQuery['tree']
        treeName = treesQuery["name_" + SUG.THR.PST["LANG"]]
    elif isx("NODE"):
        nodesQuery = database_request(
            "SELECT cluster, ! FROM nodes WHERE node = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                SUG.THR.PST["NODE"]
            ]
        )[0]
        clustersQuery = database_request(
            "SELECT tree, ! FROM clusters WHERE cluster = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                nodesQuery['cluster']
            ]
        )[0]
        treesQuery = database_request(
            "SELECT ! FROM trees WHERE tree = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                clustersQuery['tree']
            ]
        )[0]
        problemValue = None
        problemName = None
        nodeValue = SUG.THR.PST["NODE"]
        nodeName = nodesQuery["name_" + SUG.THR.PST["LANG"]]
        clusterValue = nodesQuery['cluster']
        clusterName = clustersQuery["name_" + SUG.THR.PST["LANG"]]
        treeValue = clustersQuery['tree']
        treeName = treesQuery["name_" + SUG.THR.PST["LANG"]]
    elif isx("CLUSTER"):
        clustersQuery = database_request(
            "SELECT tree, ! FROM clusters WHERE cluster = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                SUG.THR.PST["CLUSTER"]
            ]
        )[0]
        treesQuery = database_request(
            "SELECT ! FROM trees WHERE tree = ?",
            [
                "name_" + SUG.THR.PST["LANG"],
                clustersQuery['tree']
            ]
        )[0]
        problemValue = None
        problemName = None
        nodeValue = None
        nodeName = None
        clusterValue = SUG.THR.PST["CLUSTER"]
        clusterName = clustersQuery["name_" + SUG.THR.PST["LANG"]]
        treeValue = clustersQuery['tree']
        treeName = treesQuery["name_" + SUG.THR.PST["LANG"]]
    else:
        raise Error()
    result = {
        "problem": {
            "value": problemValue,
            "name": problemName
        },
        "node": {
            "value": nodeValue,
            "name": nodeName
        },
        "cluster": {
            "value": clusterValue,
            "name": clusterName
        },
        "tree": {
            "value": treeValue,
            "name": treeName
        }
    }
    return set_response(result)